src/platforms/hosted/bmda_gpiod.c

```diff
@@ -0,0 +1,264 @@
+{
+	if (pin) {
+		bmda_gpiod_debug_pin(pin, "input", false, false);
+		if (gpiod_line_set_direction_input(pin)) {
+			DEBUG_ERROR("Failed to set pin to input errno: %d", errno);
+			exit(1);
+		}
+	} else {
+		DEBUG_ERROR("BUG! attempt to set uninit GPIO to input");
+	}
+}
+
+void bmda_gpiod_mode_output(struct gpiod_line *pin)
+{
+	if (pin) {
+		bmda_gpiod_debug_pin(pin, "output", false, false);
+		if (gpiod_line_set_direction_output(pin, 0)) {
+			DEBUG_ERROR("Failed to set pin to output errno: %d", errno);
+			exit(1);
+		}
+	} else {
+		DEBUG_ERROR("BUG! attempt to set uninit GPIO to output");
+	}
+}
+
+static bool bmda_gpiod_parse_gpio(const char *name, char *gpio)
+{
+	DEBUG_INFO("GPIO set %s: %s\n", name, gpio);
+	char *offset = strstr(gpio, ":");
+	if (!offset)
+		return false;
+	*offset = '\0';
+	offset++;
+	char *end = offset + strlen(offset);
+	char *valid;
+
+	unsigned long offset_val = strtoul(offset, &valid, 10);
+
+	if (valid != end || offset_val > UINT_MAX) {
+		return false;
+	}
+
+	/* This is an unsigned int because that is what gpiod_line_get consumes. */
+	unsigned int gpio_off = (unsigned int)offset_val;
+
+	DEBUG_INFO("gpiochip: %s offset: %u\n", gpio, gpio_off);
+	struct gpiod_line *line = gpiod_line_get(gpio, gpio_off);
+	if (!line) {
+		DEBUG_ERROR("Couldn't get GPIO: %s:%u errno: %d\n", gpio, gpio_off, errno);
+		return false;
+	}
+
+	if (!strcmp("tck", name)) {
+		if (gpiod_line_request_output_flags(line, "bmda-tck", GPIOD_LINE_REQUEST_FLAG_BIAS_DISABLE, 0))
+			goto out_fail_close;
+		bmda_gpiod_tck_pin = line;
+	} else if (!strcmp("tms", name)) {
+		if (gpiod_line_request_output_flags(line, "bmda-tms", GPIOD_LINE_REQUEST_FLAG_BIAS_DISABLE, 0))
+			goto out_fail_close;
+		bmda_gpiod_tms_pin = line;
+	} else if (!strcmp("tdi", name)) {
+		if (gpiod_line_request_output_flags(line, "bmda-tdi", GPIOD_LINE_REQUEST_FLAG_BIAS_DISABLE, 0))
+			goto out_fail_close;
+		bmda_gpiod_tdi_pin = line;
+	} else if (!strcmp("tdo", name)) {
+		if (gpiod_line_request_input_flags(line, "bmda-tdo", GPIOD_LINE_REQUEST_FLAG_BIAS_DISABLE))
+			goto out_fail_close;
+		bmda_gpiod_tdo_pin = line;
+	} else if (!strcmp("swdio", name)) {
+		if (gpiod_line_request_input_flags(line, "bmda-swdio", GPIOD_LINE_REQUEST_FLAG_BIAS_DISABLE))
+			goto out_fail_close;
+		bmda_gpiod_swdio_pin = line;
+	} else if (!strcmp("swclk", name)) {
+		if (gpiod_line_request_output_flags(line, "bmda-swclk", GPIOD_LINE_REQUEST_FLAG_BIAS_DISABLE, 0))
+			goto out_fail_close;
+		bmda_gpiod_swclk_pin = line;
+	} else {
+		DEBUG_ERROR("Unrecognised signal name: %s\n", name);
+		gpiod_chip_close(gpiod_line_get_chip(line));
+		return false;
+	}
+
+	DEBUG_INFO("Line consumer: %s\n", gpiod_line_consumer(line));
+
+	return true;
+
+out_fail_close:
+	DEBUG_ERROR("Requesting gpio failed errno: %d", errno);
+	gpiod_chip_close(gpiod_line_get_chip(line));
+	return false;
+}
+
+static bool bmda_gpiod_parse_gpiomap(const char *gpio_map)
+{
+	bool ret = true;
+	char *saveptr;
```

Comment:
If possible, please initialise this and move it next to the first strtok_r() it gets used in, so we can try and prevent possible future UB read-before-init issues.

Reply:
Done